lib/step.js

```diff
@@ -38,8 +38,34 @@ class Step {
     this.metaStep = undefined;
     /** @member {string} */
     this.stack = '';
-    /** @member {number} */
-    this.totalTimeout = undefined;
+
+    const timeouts = new Map();
+    /**
+     * @method
+     * @returns {number|undefined}
+     */
+    this.getTotalTimeout = function () {
+      let totalTimeout;
+      // iterate over all timeouts starting from highest values of source
+      new Map([...timeouts.entries()].sort().reverse()).forEach((timeout, source) => {
+        if (timeout !== undefined && (
+          source !== 0 // souces > 0 always override value set earlier
+          || totalTimeout === undefined // source = 0 && totalTimeout === undefined => use timeout
+          || timeout > 0 && (timeout < totalTimeout || totalTimeout === 0) // source = 0 => timeout override smaller values or value of no timeout (totalTimeout === 0)
+        )) {
+          totalTimeout = timeout;
+        }
+      });
+      return totalTimeout;
+    };
+    /**
+     * @method
+     * @param {number} timeout - timeout in milliseconds or 0 if no timeout
+     * @param {number} source - 0 - test/suite, 1 - config force=true, 2 - code, 3 - config force=false
+     */
+    this.setTotalTimeout = function (timeout, source = 3) {
```

Comment:
`source` is not something that is really clear here.
should we call it `priority`?

Reply:
yep, we could call it priority. lower priority value gives higher priority to timeout value. I thought of calling it priority but I ended up calling it source as the logic here has very specific expectations for levels/priorities for timeout values from different sources and I tried to think of scenario where some other source of timeout try to meddle with this and couldn't.

Reply:
other plugins could also use this API to add their timeouts... So the priority is cleaner. But if lower value makes things more prioritized then let's call it `order`. In this case, `0` should be understandable to be the first value to pull out.